website/project/views/comment.py

```diff
@@ -13,106 +14,169 @@
         raise HTTPError(http.BAD_REQUEST)
     return target.referent
 
 
-def collect_discussion(target, users=None):
-
-    users = users or collections.defaultdict(list)
-    for comment in getattr(target, 'commented', []):
-        if not comment.is_deleted:
-            users[comment.user].append(comment)
-        collect_discussion(comment, users=users)
+def collect_discussion(target, users):
+    if not getattr(target, 'commented', None) is None:
+        for comment in getattr(target, 'commented', []):
+            if not (comment.is_deleted or comment.is_hidden):
+                users[comment.user].append(comment)
+            collect_discussion(comment, users=users)
     return users
 
 
-@must_be_contributor_or_public
-def comment_discussion(**kwargs):
+def comment_discussion(comments, node, anonymous=False, widget=False):
 
-    node = kwargs['node'] or kwargs['project']
-    auth = kwargs['auth']
-    users = collect_discussion(node)
-    anonymous = has_anonymous_link(node, auth)
-    # Sort users by comment frequency
-    # TODO: Allow sorting by recency, combination of frequency and recency
-    sorted_users = sorted(
+    users = collections.defaultdict(list)
+    for comment in comments:
+        if not (comment.is_deleted or comment.is_hidden):
+            users[comment.user].append(comment)
+        if not widget:
+            collect_discussion(comment, users=users)
+
+    sorted_users_frequency = sorted(
         users.keys(),
         key=lambda item: len(users[item]),
         reverse=True,
     )
 
+    def get_recency(item):
+        most_recent = users[item][0].date_created
+        for comment in users[item][1:]:
+            if comment.date_created > most_recent:
+                most_recent = comment.date_created
+        return most_recent
+
+    sorted_users_recency = sorted(
+        users.keys(),
+        key=lambda item: get_recency(item),
+        reverse=True,
+    )
+
     return {
-        'discussion': [
-            {
-                'id': privacy_info_handle(user._id, anonymous),
-                'url': privacy_info_handle(user.url, anonymous),
-                'fullname': privacy_info_handle(user.fullname, anonymous, name=True),
-                'isContributor': node.is_contributor(user),
-                'gravatarUrl': privacy_info_handle(
-                    gravatar(
-                        user, use_ssl=True, size=settings.GRAVATAR_SIZE_DISCUSSION,
-                    ),
-                    anonymous
-                ),
-
-            }
-            for user in sorted_users
+        'discussionByFrequency': [
+            serialize_discussion(node, user, len(users[user]), anonymous)
+            for user in sorted_users_frequency
+        ],
+        'discussionByRecency': [
+            serialize_discussion(node, user, len(users[user]), anonymous)
+            for user in sorted_users_recency
         ]
     }
 
 
+def serialize_discussion(node, user, num, anonymous=False):
```

Comment:
This is a misnomer; this serializes a user, not comments. 

We already have a number of user serialization function, e.g. `website.profile.util.serialize_user`. Can we reuse one of those?